### PR TITLE
feat: outbound cctx scheduling with rate limiter

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

- Added `CctxListPendingWithinRateLimit` new gRPC query for zetaclient to use for rate limiting purpose.
- Adjusted zetaclient code to switch to the new gRPC query.

Closes: [2040](https://github.com/zeta-chain/node/issues/2040)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 1]
```diff
@@ -177,3 +177,206 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
...
+	// the criteria to stop adding cctxs to the result
+	maxCCTXsReached := func() bool {
+		// #nosec G701 len always positive
+		return uint32(len(cctxs)) >= limit
+	}
+
+	// query pending cctxs for each supported chain
+	chains := k.zetaObserverKeeper.GetSupportedChains(ctx)
+ChainLoop:
+	for _, chain := range chains {
```
Comment (lumtis) [inline x/crosschain/keeper/grpc_query_cctx.go:245]: It's one chain after another? So the first chains would take everything if the limit is exceeded?

**Review by lumtis — COMMENTED** [review]
Just wondering, do we stop listing to pending cctxs outside of the window boundaries? What happen if a cctx is left behind the left boundary?

File: x/crosschain/keeper/grpc_query_cctx.go (hunk 1 above)
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go:248]: for now, yeah

**Review by brewmaster012 — APPROVED** [review]
Went over the main logic of the new RPC call that lists pending cctx with rate limiter, the logic looks good to me. Have not checked the arithmetic in the conversion and accumulation.

**Comment by brewmaster012** [comment]
Would be nice to be able to test against mainnet state in some way. Since this is read only RPC, maybe try a slightly mocked RPC and connect to mainnet node and do the query as sanity check.

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 2]
```diff
@@ -177,3 +178,260 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
 		TotalPending: totalPending,
 	}, nil
 }
+
+// CctxListPendingWithinRateLimit returns a list of pending cctxs that do not exceed the outbound rate limit
+// a limit for the number of cctxs to return can be specified or the default is MaxPendingCctxs
+func (k Keeper) CctxListPendingWithinRateLimit(c context.Context, req *types.QueryListCctxPendingWithRateLimitRequest) (*types.QueryListCctxPendingWithRateLimitResponse, error) {
```
Comment (lumtis) [inline x/crosschain/keeper/grpc_query_cctx.go:184]: Can we create a new file for the query?
`grpc_query_cctx_rate_limit.go`

Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go:172]: yeah.

File: x/crosschain/keeper/cctx_utils.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ func (k Keeper) GetRevertGasLimit(ctx sdk.Context, cctx types.CrossChainTx) (uin
 	return 0, nil
 }
 
-func IsPending(cctx types.CrossChainTx) bool {
+func IsPending(cctx *types.CrossChainTx) bool {
```
Comment (kingpinXD) [inline x/crosschain/keeper/cctx_utils.go:89]: Is this change needed ?

File: x/crosschain/keeper/grpc_query_cctx.go (hunk 2 above)
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go]: https://github.com/zeta-chain/node/pull/2045/commits/f31c988378329b4e88a85ebeee43b57990b09b92

File: x/crosschain/keeper/cctx_utils.go (hunk 3 above)
Comment (ws4charlie) [reply x/crosschain/keeper/cctx_utils.go:89]: yeah.

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 4]
```diff
@@ -161,19 +154,266 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
...
+	}
+
+	// check limit and use default MaxPendingCctxs if not specified
+	if req.Limit > MaxPendingCctxs {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("limit exceeds max limit of %d", MaxPendingCctxs))
+	}
+	limit := req.Limit
+	if limit == 0 {
+		limit = MaxPendingCctxs
+	}
```
Comment (kingpinXD) [inline x/crosschain/keeper/grpc_query_cctx.go]:
```suggestion
	if limit == 0 || req.Limit > MaxPendingCctxs{
		limit = MaxPendingCctxs
	}
```

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 5]
```diff
@@ -161,19 +154,266 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
...
+
+	// get current height and tss
+	ctx := sdk.UnwrapSDKContext(c)
+	height := ctx.BlockHeight()
+	if height <= 0 {
+		return nil, status.Error(codes.OutOfRange, "height out of range")
+	}
+	tss, found := k.zetaObserverKeeper.GetTSS(ctx)
+	if !found {
+		return nil, status.Error(codes.Internal, "tss not found")
```
Comment (kingpinXD) [inline x/crosschain/keeper/grpc_query_cctx.go]:
```suggestion
		return nil, observertypes.ErrTssNotFound
```

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 6]
```diff
@@ -161,19 +154,266 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
...
+	}
+
+	// check rate limit flags to decide if we should apply rate limit
+	applyLimit := true
+	rateLimitFlags, found := k.GetRateLimiterFlags(ctx)
+	if !found || !rateLimitFlags.Enabled {
+		applyLimit = false
+	}
```
Comment (kingpinXD) [inline x/crosschain/keeper/grpc_query_cctx.go]: It might be a good idea to add this under crosschain-flags

Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go]: I think separate flag probably works better if it's intended for a very different purpose.

File: x/crosschain/keeper/grpc_query_cctx.go [hunk 7]
```diff
@@ -161,19 +154,266 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
...
+
+// CctxListPendingWithinRateLimit returns a list of pending cctxs that do not exceed the outbound rate limit
+// a limit for the number of cctxs to return can be specified or the default is MaxPendingCctxs
+func (k Keeper) CctxListPendingWithinRateLimit(c context.Context, req *types.QueryListCctxPendingWithRateLimitRequest) (*types.QueryListCctxPendingWithRateLimitResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	// check limit and use default MaxPendingCctxs if not specified
+	if req.Limit > MaxPendingCctxs {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go]: https://github.com/zeta-chain/node/pull/2045/commits/6ad94af0ccf6399dbf75ed82b1a52178fc30ec1b

File: x/crosschain/keeper/grpc_query_cctx.go (hunk 4 above)
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx.go]: https://github.com/zeta-chain/node/pull/2045/commits/6ad94af0ccf6399dbf75ed82b1a52178fc30ec1b

**Review by lumtis — APPROVED** [review]

**Comment by ws4charlie** [comment]
> Just wondering, do we stop listing to pending cctxs outside of the window boundaries? What happen if a cctx is left behind the left boundary?

When looping backwards, the query picks up pending cctx in the range `[nonceLow-1000, nonceLow]`. If the cctx on `nonceLow` hasn't hit the left boundary (e.g,, current height - 2hours_of_blocks), the backwards loop will continue until it hits the left boundary in order to calculate the total withdrawn amount during the rate limiter window.

**Review by kingpinXD — APPROVED** [review]
